exec/cnex/lib/os_posix.c

```diff
@@ -0,0 +1,192 @@
+// The following is needed for kill() on Posix platforms.
+#define _POSIX_C_SOURCE 200809L
+
+#include <assert.h>
+#include <errno.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <sys/param.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+
+#include "cell.h"
+#include "enums.h"
+#include "exec.h"
+#include "number.h"
+#include "object.h"
+#include "stack.h"
+
+
+
+static Array *g_children;
+
+typedef struct tagTProcessHandle {
+    pid_t proc;
+} TProcessHandle;
+
+
+
+static TProcessHandle *check_process(TExecutor *exec, Object *pp)
+{
+    if (pp == NULL || pp->ptr == NULL || ((TProcessHandle*)pp->ptr)->proc == 0) {
+        exec->rtl_raise(exec, "OsException.InvalidProcess", "", BID_ZERO);
+        return NULL;
+    }
+    return pp->ptr;
+}
+
+void object_releaseProcessObject(Object *o)
+{
+    if (o != NULL) {
+        assert(o->refcount > 0);
+        o->refcount--;
+
+        if (o->refcount <= 0) {
+            if (o->ptr != NULL) {
+                free(o->ptr);
+            }
+            free(o);
+        }
+    }
+}
+
+Cell *object_processObjectToString(Object *self)
+{
+    Cell *r = cell_fromCString("<PROCESS ");
+    char s[16];
+    snprintf(s, 16, "%d>", ((TProcessHandle*)self->ptr)->proc);
+    r->string = string_appendCString(r->string, s);
+    return r;
+}
+
+Object *object_createProcessObject(void *p)
+{
+    Object *r = object_createObject();
+    r->ptr = p;
+
+    r->release = object_releaseProcessObject;
+    r->toString = object_processObjectToString;
+    return r;
+}
+
+void closer()
+{
+    if (g_children == NULL  || g_children->data == NULL) {
+        return;
+    }
+    for (size_t i = 0; i < g_children->size; i++) {
+        if (g_children->data[i].type != cNothing) {
+            TProcessHandle *p = g_children->data[i].object->ptr;
+            kill(-(p->proc), 9);
+        }
+    }
+}
+
+
+
+void os_initModule()
+{
+    g_children = array_createArray();
+}
+
+void os_shutdownModule()
+{
+    array_freeArray(g_children);
+    g_children = NULL;
+}
+
+
+void os_chdir(TExecutor *exec)
+{
+    char *path = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    int r = chdir(path);
+    if (r != 0) {
+        exec->rtl_raise(exec, "OsException.PathNotFound", path, BID_ZERO);
+    }
+    free(path);
+}
+
+void os_getcwd(TExecutor *exec)
+{
+    char buf[MAXPATHLEN];
+    getcwd(buf, sizeof(buf));
+
+    push(exec->stack, cell_fromString(string_createCString(buf)));
+}
+
+void os_kill(TExecutor *exec)
+{
+    Object *process = top(exec->stack)->object; pop(exec->stack);
+    TProcessHandle *p = check_process(exec, process);
+
+    kill(-(pid_t)p->proc, SIGTERM);
```

Comment:
Why is there a `(pid_t)` cast here? It's already the right type.

Reply:
Removed cast.